hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/FileGroupReaderBasedMergeHandle.java

```diff
@@ -267,26 +393,141 @@ private static class CDCCallback<T> implements BaseFileUpdateCallback<T> {
+    public void onUpdate(String recordKey, BufferedRecord<T> previousRecord, BufferedRecord<T> mergedRecord) {
       cdcLogger.put(recordKey, convertOutput(previousRecord), Option.of(convertOutput(mergedRecord)));
-
     }
 
     @Override
-    public void onInsert(String recordKey, T newRecord) {
+    public void onInsert(String recordKey, BufferedRecord<T> newRecord) {
       cdcLogger.put(recordKey, null, Option.of(convertOutput(newRecord)));
-
     }
 
     @Override
-    public void onDelete(String recordKey, T previousRecord) {
+    public void onDelete(String recordKey, BufferedRecord<T> previousRecord, HoodieOperation hoodieOperation) {
       cdcLogger.put(recordKey, convertOutput(previousRecord), Option.empty());
-
     }
 
-    private GenericRecord convertOutput(T record) {
-      T convertedRecord = outputConverter.get().map(converter -> record == null ? null : converter.apply(record)).orElse(record);
+    private GenericRecord convertOutput(BufferedRecord<T> record) {
+      T convertedRecord = outputConverter.get().map(converter -> record == null ? null : converter.apply(record.getRecord())).orElse(record.getRecord());
       return convertedRecord == null ? null : readerContext.getRecordContext().convertToAvroRecord(convertedRecord, requestedSchema.get());
     }
   }
+
+  private static class RecordLevelIndexCallback<T> implements BaseFileUpdateCallback<T> {
+    private final WriteStatus writeStatus;
+    private final HoodieRecordLocation fileRecordLocation;
+    private final String partitionPath;
+
+    public RecordLevelIndexCallback(WriteStatus writeStatus, HoodieRecordLocation fileRecordLocation, String partitionPath) {
+      this.writeStatus = writeStatus;
+      this.fileRecordLocation = fileRecordLocation;
+      this.partitionPath = partitionPath;
+    }
+
+    @Override
+    public void onUpdate(String recordKey, BufferedRecord<T> previousRecord, BufferedRecord<T> mergedRecord) {
+      writeStatus.addRecordDelegate(HoodieRecordDelegate.create(recordKey, partitionPath, fileRecordLocation, fileRecordLocation, mergedRecord.getHoodieOperation() == HoodieOperation.UPDATE_BEFORE));
+    }
+
+    @Override
+    public void onInsert(String recordKey, BufferedRecord<T> newRecord) {
+      writeStatus.addRecordDelegate(HoodieRecordDelegate.create(recordKey, partitionPath, null, fileRecordLocation, newRecord.getHoodieOperation() == HoodieOperation.UPDATE_BEFORE));
+    }
+
+    @Override
+    public void onDelete(String recordKey, BufferedRecord<T> previousRecord, HoodieOperation hoodieOperation) {
+      // The update before operation is used when a deletion is being sent to the old File Group in a different partition.
+      // In this case, we do not want to delete the record metadata from the index.
+      writeStatus.addRecordDelegate(HoodieRecordDelegate.create(recordKey, partitionPath, fileRecordLocation, null, hoodieOperation == HoodieOperation.UPDATE_BEFORE));
```

Comment:
`hoodieOperation == HoodieOperation.UPDATE_BEFORE` is always false.

Reply:
I simplified this in a recent commit